plat/kvm/x86/multiboot2.c

```diff
@@ -0,0 +1,180 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2024, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/essentials.h>
+#include <uk/arch/limits.h>
+#include <uk/arch/types.h>
+#include <uk/arch/paging.h>
+#include <uk/plat/bootstrap.h>
+#include <uk/plat/common/bootinfo.h>
+#include <uk/plat/common/lcpu.h>
+#include <uk/plat/common/memory.h>
+#include <uk/plat/common/sections.h>
+#include <uk/reloc.h>
+#include <kvm-x86/multiboot2.h>
+
+#include <errno.h>
+#include <string.h>
+
+#define multiboot2_crash(msg, rc)	ukplat_crash()
+
+/* Round up 'x' to the nearest multiple of 'MULTIBOOT_TAG_ALIGN' */
+#define ROUND_UP_TO_ALIGNMENT(x) \
+	(((x) + (MULTIBOOT_TAG_ALIGN - 1)) & ~(MULTIBOOT_TAG_ALIGN - 1))
+
+void _ukplat_entry(struct lcpu *lcpu, struct ukplat_bootinfo *bi);
+
+static inline void mrd_insert(struct ukplat_bootinfo *bi,
+			      const struct ukplat_memregion_desc *mrd)
+{
+	int rc;
+
+	if (unlikely(mrd->len == 0))
+		return;
+
+	rc = ukplat_memregion_list_insert(&bi->mrds, mrd);
+	if (unlikely(rc < 0))
+		multiboot2_crash("Cannot insert bootinfo memory region", rc);
+}
+
+/**
+ * Multiboot entry point called after lcpu initialization. We enter with the
+ * 1:1 boot page table set. Physical and virtual addresses thus match for all
+ * regions in the mapped range.
+ */
+void multiboot2_entry(struct lcpu *lcpu, unsigned long addr)
+{
+	struct ukplat_bootinfo *bi;
```

Comment:
I have a personal preference when it comes to variable declaration. I will only mention it once, feel free to disregard/apply it if you also dislike/like it. I generally declare these in reverse xmas tree order...
e.g. https://github.com/unikraft/unikraft/blob/staging/plat/kvm/efi.c#L299
Or pretty much how that whole file does it. There are many other files that follow these patterns. Though this is not mandated by our coding style (though I hope it will) and others do not really use this pattern, I personally like it better.